torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+
+    if isinstance(current_scale, float):
+        current_scale = torch.full((1,), current_scale, dtype=torch.float32, device="cuda")
+
+    # Eventually, we'd like this to become a multi-tensor apply call dispatched via NestedTensor.
+    for group in self.param_groups:
+        for param in group["params"]:
+            if param.grad is not None:
+                if param.grad.dtype == torch.float16:
+                    raise ValueError("Attempting to unscale FP16 gradients.  If you want to check for infs/nans  "
+                                     "without unscaling, use optimizer.check_infs() instead.")
+                else:
+                    torch._amp_unscale_inf_check_(param.grad, current_scale, found_inf)
+
+    if scale_scheduler is None:
+        next_recommended_scale = _next_recommended_scale(current_scale, found_inf,
+                                                         scale_growth_factor, scale_backoff_factor)
+    else:
+        next_recommended_scale = scale_scheduler(current_scale, found_inf)
+
+    return found_inf, next_recommended_scale
+
+
+def check_inf(self, scaling_enabled=True):
+    r"""
+    :func:`check_inf` is patched onto an optimizer instance by :func:`add_amp_attributes`.  Since :func:`check_inf`
+    becomes a method, ``self`` is passed automatically.  :func:`check_inf` should only be called as an attribute of
+    its optimizer instance (e.g. ``opt.check_inf()``), not as a free function.
+
+    Check the gradients currently owned by this optimizer for infs/nans, without unscaling.  For typical
+    implementations of gradient unscaling in your script, it's not necessary to call this function; prefer
+    :func:`unscale_and_step` or the :func:`unscale`+:func:`step_after_unscale` combination as shown in the Examples.
+
+    Arguments:
+        scaling_enabled (bool, optional):  If ``False``, :func:`check_inf` becomes a no-op and returns ``None``.
+
+    Returns:
+        ``found_inf``, a ``torch.cuda.FloatTensor`` that contains > 0 if an inf/nan is present in the current gradients,
+        and 0.0 otherwise.  If ``scaling_enabled=False``, the return value will be ``None``.
+    """
+    if not scaling_enabled:
+        return None
+
+    found_inf = torch.full((1,), 0.0, dtype=torch.float32, device="cuda")
+
+    # This duplicates code from unscale.  In principle I could just call into unscale(1.0), but that would error
+    # if gradients are fp16, and I want check_inf to be usable on fp16 gradients.  I also don't want to add an argument
+    # to "unscale" that permits unscaling in fp16.
+    for group in self.param_groups:
+        for param in group["params"]:
+            if param.grad is not None:
+                torch._amp_unscale_inf_check_(param.grad,
+                                              torch.full((1,), 1.0, dtype=torch.float32, device="cuda"),
+                                              found_inf)
+
+    return found_inf
+
+
+def step_after_unscale(self,
```

Comment:
Do we need this function? Isn't it OK for the user to just call step() in this case?